### PR TITLE
feat: `is_allocated()`

### LINKED ISSUE #25: [Feature]: Allocation API proposal
The following API changes would make sense to me after using this:

1. Rename `Ledger::add()` to `Ledger::map()`.
2. Rename `Ledger::delete()` to `Ledger::unmap()`.
3. Change API from `Ledger::add(record)` to `Ledger::add(region, access)`.

These are descriptive because their purpose is to intervene `mmap()` and `munmap()`.

3rd proposal comes from practical use. It's very impractical to always have to construct a one-shot instance of `Record` only to make a one API call, and then throw it away.

### PR DESCRIPTION
**Pull request by jarkkojs**
In addition: rename `add()` to `allocate()`

Closes: #25
Signed-off-by: Jarkko Sakkinen <jarkko@profian.com>

### CODE REVIEW

**Comment by jarkkojs** [comment]
Added unit tests.

**Comment by jarkkojs** [comment]
@npmccallum, I need this for heap. Even if there is a proper find mechanism in place, this function benefits from its own implementation given that it does not care about order. This works for my use case perfectly. And in the API it is convenient to have both `"is" and "where" queries.

File: src/lib.rs [hunk 1]
```diff
@@ -1,6 +1,6 @@
 //! A ledger for memory mappings.
 
-#![cfg_attr(not(test), no_std)]
+//#![cfg_attr(not(test), no_std)]
```
Comment (npmccallum) [inline src/lib.rs]: Don't. :)

File: src/lib.rs [hunk 2]
```diff
@@ -216,6 +217,34 @@ impl<const N: usize> Ledger<N> {
...
+
+            if region.start > middle_region.start {
+                start = middle + 1;
+            } else {
+                end = middle - 1;
+            }
+        }
+
+        false
+    }
```
Comment (npmccallum) [inline src/lib.rs:268]: Rust already offers methods for binary search on slices.

**Review by npmccallum — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 1 above)
Comment (jarkkojs) [reply src/lib.rs]: a mistake (had some temporary prints). thanks for spotting that out and apologies.

**Comment by jarkkojs** [comment]
@npmccallum issues resolved

**Comment by jarkkojs** [comment]
Updated to use `is_ok_or_err()`.

**Comment by jarkkojs** [comment]
Updated to `unwrap_or_else()`.

**Comment by jarkkojs** [comment]
clean up. I think it is now quite polished. @npmccallum

**Review by npmccallum — CHANGES_REQUESTED** [review]
Please separate `.is_allocated()` from the rename stuff. They are unrelated.

**Comment by jarkkojs** [comment]
Usage example:
```rust
    fn is_allocated(&self, index: usize) -> bool {
        self.is_allocated_range(index..index)
    }

    fn is_allocated_range(&self, range: core::ops::Range<usize>) -> bool {
        self.ledger.is_allocated(Region::new(
            Address::new(range.start * size_of::<Page>()),
            Address::new(range.end * size_of::<Page>()),
        ))
    }
```

**Comment by jarkkojs** [comment]
@npmccallum now it is only `is_allocated()`, and also separate issue created.

**Comment by jarkkojs** [comment]
just rebased to latest mainline

**Comment by jarkkojs** [comment]
not a priority, I can do this internally for enarx for now (and not aligned what I said of not adding seek's before enarx is done).

**Pull request closed without merging**